### PR TITLE
CT-3704 address contact book

### PR DESCRIPTION
**Pull request by willmcb**
## Description
Adds a CRUD Address book to the tool for Branston users

- Note the model has an email field as well as this is anticipated as an addition to come soon.

Questions for reviewers:
- Is it work adding a postcode validation?
- does it make sense to add the name to the start of the address on the index page?
- do you think the form needs sub-headings for the fields?

## Self-review checklist

* [ ] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [ ] (3) Tests passing
* [ ] (4) Branch ready to be merged (not work in progress)
* [ ] (5) No superfluous changes in diff
* [ ] (6) No TODO's without new ticket numbers
* [ ] (7) PR Prefixed with ticket number e.g. `CT-7654 ...`

### Screenshots
Index: 
Screenshot 2021-07-30 at 15 23 51

Edit: 
Screenshot 2021-07-30 at 15 24 31

New:
Screenshot 2021-07-30 at 15 25 02

### Related JIRA tickets
[CT-3704](https://dsdmoj.atlassian.net/browse/CT-3704)

### Deployment
- should I add a feature flag to this, as the MVP relies on the copy functionality as outlined in CT-3703?

### Manual testing instructions
- login as branston user
- click addresses
- CRUD an address.

### CODE REVIEW

File: app/views/contacts/index.html.slim [hunk 1]
```diff
@@ -0,0 +1,29 @@
+- content_for :heading
+  = "Address book"
+
+= render partial: 'layouts/header'
+
+.grid-row
+  .column-full.button-holder
+    = link_to 'Add new address', new_contact_path, class: 'button'
+
+table
```
Comment (digitalali-moj) [inline app/views/contacts/index.html.slim:10]: Please check if mobile works, if not, see if adding class "report" helps to table element.

File: app/views/contacts/_form.html.slim [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      = f.text_field :address_line_2
+      = f.text_field :town
+      = f.text_field :county
+      = f.text_field :postcode
+      = f.radio_button_fieldset :contact_type, choices: Contact::contact_types.keys, inline: true
+
+  .button-holder
+    = f.submit 'Submit', class: 'button'
+
+  = link_to('Cancel', contacts_url)
```
Comment (digitalali-moj) [inline app/views/contacts/_form.html.slim:15]: I think this should be in button-holder

File: app/views/contacts/edit.html.slim [hunk 3]
```diff
@@ -0,0 +1,13 @@
+= link_to("Back", contacts_url, class: 'govuk-back-link')
+- content_for :heading
+  = "Edit address"
+
+= render partial: 'layouts/header'
+
+= GovukElementsErrorsHelper.error_summary @contact,
+  "#{pluralize(@contact.errors.count, t('common.error'))} #{ t('common.summary_error')}", ""
+
+.form-group
```
Comment (digitalali-moj) [inline app/views/contacts/edit.html.slim:10]: I think we normally use this for form fields, which is autogenerated, but in some places it's before the form. But I'm not sure if we need it.

File: app/views/contacts/_form.html.slim [hunk 4]
```diff
@@ -0,0 +1,15 @@
+= form_for @contact do |f|
+
+  .form-group
```
Comment (digitalali-moj) [inline app/views/contacts/_form.html.slim:3]: I'm not sure again if we need this. Normally the title pushes stuff down if it's for gaps. I guess I'm just asking to try without.

File: app/views/contacts/_form.html.slim [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+      = f.text_field :name
+      = f.text_field :address_line_1
+      = f.text_field :address_line_2
+      = f.text_field :town
+      = f.text_field :county
+      = f.text_field :postcode
+      = f.radio_button_fieldset :contact_type, choices: Contact::contact_types.keys, inline: true
```
Comment (digitalali-moj) [inline app/views/contacts/_form.html.slim:10]: Is there something different about this as normally the radios come out in separate lines?

**Review by digitalali-moj — COMMENTED** [review]
Probably giving feedback too early, but just some minor f/e comments. :)

**Comment by Polsonby** [comment]
This looks amazing, @willmcb will review but my £0.02 on your questions:
"is it worth adding a postcode validation?" - Not in my opinion, no
"does it make sense to add the name to the start of the address on the index page?" I think it's better without it repeated on the index page, but when the address is used for real it should be concatenated on the front (I know, not this ticket!)
"do you think the form needs sub-headings for the fields?" - I don't think so...

File: app/models/contact.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    format_address("\n")
+  end
+
+  def inline_address
+    format_address(", ")
+  end
+
+  private
+
+  def format_address(seperator)
```
Comment (Polsonby) [inline app/models/contact.rb:19]: Tiny typo? 

```suggestion
  def format_address(separator)
```

File: app/models/contact.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      name,
+      address_line_1,
+      address_line_2,
+      town,
+      county,
+      postcode
+    ].compact
+     .reject(&:empty?)
+     .join(seperator)
+  end
```
Comment (Polsonby) [inline app/models/contact.rb:30]: nice :)

File: app/models/contact.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    [
+      name,
+      address_line_1,
+      address_line_2,
+      town,
+      county,
+      postcode
+    ].compact
+     .reject(&:empty?)
+     .join(seperator)
```
Comment (Polsonby) [inline app/models/contact.rb:29]:
```suggestion
     .join(separator)
```

File: config/settings.yml [hunk 9]
```diff
@@ -164,6 +164,11 @@ global_navigation:
       visibility:
         - admin
 
+    addresses:
+      path: /contacts
+      visibility:
+        - BRANSTON
+
```
Comment (Polsonby) [inline config/settings.yml:171]: Oh my goodness - that was so much simpler than what I was trying to do 🤦 completely forgot we had this bit!

**Review by Polsonby — APPROVED** [review]
Really nice work, Will, and completed in record time 😄  - be great if you had a moment to fix tiny typo but this is fab, works great  - thanks so much for taking this over.

File: app/views/contacts/index.html.slim (hunk 1 above)
Comment (willmcb) [reply app/views/contacts/index.html.slim:10]: That's brilliant thanks Jav looks much better on mobile now as it collapses to a nicely formatted table.

File: app/views/contacts/_form.html.slim (hunk 2 above)
Comment (willmcb) [reply app/views/contacts/_form.html.slim:15]: 👍

File: app/views/contacts/_form.html.slim (hunk 4 above)
Comment (willmcb) [reply app/views/contacts/_form.html.slim:3]: Nice, thanks looks fine without so I'll take it out.

File: app/views/contacts/_form.html.slim (hunk 5 above)
Comment (willmcb) [reply app/views/contacts/_form.html.slim:10]: Hmm 🤔 - yeah they do don't they? I'm not sure tbh I just copied from another form. I'll have a look and check if there I can get the default behaviour instead.

File: app/models/contact.rb (hunk 6 above)
Comment (willmcb) [reply app/models/contact.rb:19]: 🤦 thanks!

File: config/settings.yml (hunk 9 above)
Comment (willmcb) [reply config/settings.yml:171]: Yeah I was looking through the code and I stumbled upon this settings file so I thought, why not give it a try! :)

File: app/views/contacts/_form.html.slim (hunk 5 above)
Comment (willmcb) [reply app/views/contacts/_form.html.slim]: @digitalali-moj it was the `inline: true` option - not sure why I left this on! removed now and it's in the usual style.

**Pull request merged**